File: algorithms/cpp/031_next_permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if (nums.size()<2) return;
        int index=0;
        for (int i=1;i<nums.size();i++) {
            if (nums[i]>nums[i-1]) index=i;
        }
        int index2=index;
        for (int i=index;i<nums.size();i++) {
            if (nums[i]>nums[index-1]) index2=i;}
        if (index==0) swap(nums,0,nums.size()-1);
        else {
            int temp=nums[index-1];
            nums[index-1]=nums[index2];
            nums[index2]=temp;
            swap(nums,index,nums.size()-1);
        }
    }
    void swap(vector<int>& nums, int left, int right) {
        if (left>=right) return;
        for (int i=left;i<(right+left+1)/2;i++) {
            int temp=nums[i];
            nums[i]=nums[right-i+left];
            nums[right-i+left]=temp;
        }
    }
};

File: algorithms/python3/031_next_permutation.py
class Solution:
    def nextPermutation(self, nums):
        """
        Do not return anything, modify nums in-place instead.
        """
        for i in range(len(nums)-1,0,-1):
            if nums[i]>nums[i-1]:
                ind=i
                while ind<len(nums):
                    if nums[ind]>nums[i-1]:
                        ind+=1
                    else:
                        break
                ind-=1
                nums[i-1],nums[ind]=nums[ind],nums[i-1]
                for j in range(i,i+(len(nums)-i)//2):
                    nums[j],nums[len(nums)-(j-i)-1]=nums[len(nums)-(j-i)-1],nums[j]
                return
        nums.sort()


nums=[1,2,3]

a=Solution()

a.nextPermutation(nums)

